pkg/hub/controllers/finalizercontroller/finalize_controller.go

```diff
@@ -0,0 +1,249 @@
+package finalizercontroller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	clusterv1listers "github.com/open-cluster-management/api/client/cluster/listers/cluster/v1"
+	workv1client "github.com/open-cluster-management/api/client/work/clientset/versioned/typed/work/v1"
+	worklister "github.com/open-cluster-management/api/client/work/listers/work/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	rbacv1informers "k8s.io/client-go/informers/rbac/v1"
+	rbacv1client "k8s.io/client-go/kubernetes/typed/rbac/v1"
+	rbacv1listers "k8s.io/client-go/listers/rbac/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+// FinalizeController ensures manifestworks for a certain spoke cluster are deleted before
+// role/rolebinding for work agent in cluster namespace are deleted. This happens when
+// 1). the spoke cluster is deleting;
+// 2). and the role/rolebinding for work agent in the spoke cluster namespace with the
+//     required finalizer is deleting as well.
+type FinalizeController struct {
+	roleLister         rbacv1listers.RoleLister
+	roleBindingLister  rbacv1listers.RoleBindingLister
+	rbacClient         rbacv1client.RbacV1Interface
+	clusterLister      clusterv1listers.SpokeClusterLister
+	manifestWorkClient workv1client.WorkV1Interface
+	manifestWorkLister worklister.ManifestWorkLister
+}
+
+func NewFinalizeController(
+	recorder events.Recorder,
+	roleInformer rbacv1informers.RoleInformer,
+	roleBindingInformer rbacv1informers.RoleBindingInformer,
+	clusterLister clusterv1listers.SpokeClusterLister,
+	manifestWorkClient workv1client.WorkV1Interface,
+	manifestWorkLister worklister.ManifestWorkLister,
+	rbacClient rbacv1client.RbacV1Interface,
+) factory.Controller {
+
+	controller := &FinalizeController{
+		roleLister:         roleInformer.Lister(),
+		roleBindingLister:  roleBindingInformer.Lister(),
+		clusterLister:      clusterLister,
+		manifestWorkClient: manifestWorkClient,
+		manifestWorkLister: manifestWorkLister,
+		rbacClient:         rbacClient,
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return key
+		}, roleInformer.Informer(), roleBindingInformer.Informer()).
+		WithSync(controller.sync).ToController("FinalizeController", recorder)
+}
+
+func (m *FinalizeController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
```

Comment:
I see two primary cases.
1. client deletes role or rolebinding, but does not delete the namespace.  In this case, allow them to be deleted.  The controller should recreate it.
2. client deletes the namespace, which causes the roles and rolebindings to be deleted.  In this case, keep the role and rolebinding until the manifestworks are gone.

If we do it like that, I think the logic here needs some updates so that a delete of replaceable roles and rolebindings does not result in deletes of manifestworks.

I think you should also consider how orphaning a cluster would work.  I suspect it involves first removing the spoke controllers, but it's a think to consider.

Reply:
yes, I did not consider that client delete role/rolebinding. I think we can only check if all manifestwork is deleted then remove finalizer, instead of delete manifestwork implicitly. Namespace deletion will trigger manifestwork deletion. Does it make sense?

Reply:
> I see two primary cases.

I'm +1 on what David articulated here and on outlining how to orphan a cluster.